{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 1: Develop a machine learning method to identify RNA modifications from direct RNA-Seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a computational method that predicts m6A RNA modification from direct RNA-Seq data. The method should be able to train a new new model, and make predictions on unseen test data. Specifically, your method should fullfil the following requirements:\n",
    "\n",
    "Your method should contain two scripts, one for model training, and one for making predictions. The prediction script will be evaluated by other students."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Reading the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Student_evaluation_guideline.html',\n",
       " 'dataset0.json.gz',\n",
       " 'data.info.labelled',\n",
       " 'handout_TeamProject_RNAModifications.html']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(\"data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I'm putting these files into my ~/data, if it's different for u, change the path accordingly :)\n",
    "M6A_FILE_PATH = \"data/data.info.labelled\"\n",
    "DIRECT_RNA_SEQ_DATA_FILE_PATH = \"data/dataset0.json.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read m6a labels\n",
    "def read_m6A_labels(m6a_file_path):\n",
    "    m6a_df = pd.read_csv(m6a_file_path, sep=\",\")\n",
    "    m6a_df.columns = [\"gene_id\", \"transcript_id\", \"transcript_position\", \"label\"]\n",
    "    return m6a_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = read_m6A_labels(M6A_FILE_PATH)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_direct_rna_seq_data(data_path):\n",
    "    data = []\n",
    "    with gzip.open(data_path, 'rt') as f:\n",
    "        for line in f:\n",
    "            line_data = json.loads(line)\n",
    "            for transcript_id, position_data in line_data.items():\n",
    "                for transcript_position, combined_nucleotides_data in position_data.items():\n",
    "                    for combined_nucleotide, reads in combined_nucleotides_data.items():\n",
    "                        for read_idx, read in enumerate(reads):\n",
    "                            data.append({\n",
    "                                'transcript_id': transcript_id,\n",
    "                                'position': int(transcript_position),\n",
    "                                'read_id': read_idx,\n",
    "                                'x_1': read[0],\n",
    "                                'x_2': read[1],\n",
    "                                'x_3': read[2],\n",
    "                                'x_4': read[3],\n",
    "                                'x_5': read[4],\n",
    "                                'x_6': read[5],\n",
    "                                'x_7': read[6],\n",
    "                                'x_8': read[7],\n",
    "                                'x_9': read[8]\n",
    "                            })\n",
    "\n",
    "    df = pd.DataFrame(data)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq_data_df = read_direct_rna_seq_data(DIRECT_RNA_SEQ_DATA_FILE_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>read_id</th>\n",
       "      <th>x_1</th>\n",
       "      <th>x_2</th>\n",
       "      <th>x_3</th>\n",
       "      <th>x_4</th>\n",
       "      <th>x_5</th>\n",
       "      <th>x_6</th>\n",
       "      <th>x_7</th>\n",
       "      <th>x_8</th>\n",
       "      <th>x_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>3</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>4</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>95</td>\n",
       "      <td>0.00590</td>\n",
       "      <td>5.57</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>11.20</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>9.24</td>\n",
       "      <td>87.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>96</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>3.73</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.02520</td>\n",
       "      <td>14.40</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.00510</td>\n",
       "      <td>4.16</td>\n",
       "      <td>81.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>97</td>\n",
       "      <td>0.01350</td>\n",
       "      <td>4.09</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.00540</td>\n",
       "      <td>5.71</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00396</td>\n",
       "      <td>3.48</td>\n",
       "      <td>81.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>98</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>4.17</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.00973</td>\n",
       "      <td>5.68</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>1.60</td>\n",
       "      <td>81.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>99</td>\n",
       "      <td>0.00640</td>\n",
       "      <td>8.50</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>14.90</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>3.05</td>\n",
       "      <td>76.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  position  read_id      x_1   x_2    x_3      x_4    x_5  \\\n",
       "0   ENST00000000233       244        0  0.00299  2.06  125.0  0.01770  10.40   \n",
       "1   ENST00000000233       244        1  0.00631  2.53  125.0  0.00844   4.67   \n",
       "2   ENST00000000233       244        2  0.00465  3.92  109.0  0.01360  12.00   \n",
       "3   ENST00000000233       244        3  0.00398  2.06  125.0  0.00830   5.01   \n",
       "4   ENST00000000233       244        4  0.00664  2.92  120.0  0.00266   3.94   \n",
       "..              ...       ...      ...      ...   ...    ...      ...    ...   \n",
       "95  ENST00000000233       244       95  0.00590  5.57  126.0  0.01200  11.20   \n",
       "96  ENST00000000233       244       96  0.01200  3.73  124.0  0.02520  14.40   \n",
       "97  ENST00000000233       244       97  0.01350  4.09  126.0  0.00540   5.71   \n",
       "98  ENST00000000233       244       98  0.00830  4.17  121.0  0.00973   5.68   \n",
       "99  ENST00000000233       244       99  0.00640  8.50  122.0  0.00531  14.90   \n",
       "\n",
       "      x_6      x_7    x_8   x_9  \n",
       "0   122.0  0.00930  10.90  84.1  \n",
       "1   126.0  0.01030   6.30  80.9  \n",
       "2   124.0  0.00498   2.13  79.6  \n",
       "3   130.0  0.00498   3.78  80.4  \n",
       "4   129.0  0.01300   7.15  82.2  \n",
       "..    ...      ...    ...   ...  \n",
       "95  127.0  0.00564   9.24  87.3  \n",
       "96  123.0  0.00510   4.16  81.2  \n",
       "97  127.0  0.00396   3.48  81.4  \n",
       "98  124.0  0.00316   1.60  81.8  \n",
       "99  117.0  0.01030   3.05  76.1  \n",
       "\n",
       "[100 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rna_seq_data.head()\n",
    "rna_seq_data_df.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Strategy 1: Get Aggregate of all the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>x_1</th>\n",
       "      <th>x_2</th>\n",
       "      <th>x_3</th>\n",
       "      <th>x_4</th>\n",
       "      <th>x_5</th>\n",
       "      <th>x_6</th>\n",
       "      <th>x_7</th>\n",
       "      <th>x_8</th>\n",
       "      <th>x_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>2242</td>\n",
       "      <td>0.008410</td>\n",
       "      <td>3.393768</td>\n",
       "      <td>104.312077</td>\n",
       "      <td>0.009123</td>\n",
       "      <td>3.659807</td>\n",
       "      <td>99.084541</td>\n",
       "      <td>0.007017</td>\n",
       "      <td>2.583527</td>\n",
       "      <td>88.815942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>35</td>\n",
       "      <td>0.013804</td>\n",
       "      <td>7.024259</td>\n",
       "      <td>116.333333</td>\n",
       "      <td>0.009167</td>\n",
       "      <td>7.397778</td>\n",
       "      <td>119.327778</td>\n",
       "      <td>0.006937</td>\n",
       "      <td>2.754444</td>\n",
       "      <td>83.496296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>54</td>\n",
       "      <td>0.013525</td>\n",
       "      <td>3.403200</td>\n",
       "      <td>119.960000</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>6.730200</td>\n",
       "      <td>121.180000</td>\n",
       "      <td>0.006244</td>\n",
       "      <td>3.840600</td>\n",
       "      <td>83.824000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>207</td>\n",
       "      <td>0.006547</td>\n",
       "      <td>3.563962</td>\n",
       "      <td>101.907547</td>\n",
       "      <td>0.005667</td>\n",
       "      <td>7.683396</td>\n",
       "      <td>116.830189</td>\n",
       "      <td>0.006573</td>\n",
       "      <td>4.287358</td>\n",
       "      <td>77.590566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>367</td>\n",
       "      <td>0.010143</td>\n",
       "      <td>3.926346</td>\n",
       "      <td>116.057692</td>\n",
       "      <td>0.007552</td>\n",
       "      <td>8.440962</td>\n",
       "      <td>116.134615</td>\n",
       "      <td>0.006806</td>\n",
       "      <td>2.879038</td>\n",
       "      <td>78.242308</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  position       x_1       x_2         x_3       x_4  \\\n",
       "0   ENST00000000233       244  0.008264  4.223784  123.702703  0.009373   \n",
       "1   ENST00000000233       261  0.006609  3.216424  109.681395  0.006813   \n",
       "2   ENST00000000233       316  0.007570  2.940541  105.475676  0.007416   \n",
       "3   ENST00000000233       332  0.010620  6.476350  129.355000  0.008632   \n",
       "4   ENST00000000233       368  0.010701  6.415051  117.924242  0.011479   \n",
       "..              ...       ...       ...       ...         ...       ...   \n",
       "95  ENST00000001008      2242  0.008410  3.393768  104.312077  0.009123   \n",
       "96  ENST00000002165        35  0.013804  7.024259  116.333333  0.009167   \n",
       "97  ENST00000002165        54  0.013525  3.403200  119.960000  0.008390   \n",
       "98  ENST00000002165       207  0.006547  3.563962  101.907547  0.005667   \n",
       "99  ENST00000002165       367  0.010143  3.926346  116.057692  0.007552   \n",
       "\n",
       "         x_5         x_6       x_7       x_8        x_9  \n",
       "0   7.382162  125.913514  0.007345  4.386989  80.570270  \n",
       "1   3.226535  107.889535  0.007710  3.016599  94.290698  \n",
       "2   3.642703   98.947027  0.007555  2.087146  89.364324  \n",
       "3   2.899200   97.836500  0.006101  2.236520  89.154000  \n",
       "4   5.870303  121.954545  0.010019  4.260253  85.178788  \n",
       "..       ...         ...       ...       ...        ...  \n",
       "95  3.659807   99.084541  0.007017  2.583527  88.815942  \n",
       "96  7.397778  119.327778  0.006937  2.754444  83.496296  \n",
       "97  6.730200  121.180000  0.006244  3.840600  83.824000  \n",
       "98  7.683396  116.830189  0.006573  4.287358  77.590566  \n",
       "99  8.440962  116.134615  0.006806  2.879038  78.242308  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregated_df = rna_seq_data_df.drop(columns=['read_id']).groupby(['transcript_id', 'position']).mean().reset_index()\n",
    "aggregated_df.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Train Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
