{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 1: Develop a machine learning method to identify RNA modifications from direct RNA-Seq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a computational method that predicts m6A RNA modification from direct RNA-Seq data. The method should be able to train a new new model, and make predictions on unseen test data. Specifically, your method should fullfil the following requirements:\n",
    "\n",
    "Your method should contain two scripts, one for model training, and one for making predictions. The prediction script will be evaluated by other students."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, classification_report, balanced_accuracy_score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Read M6A Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data.info.labelled', 'dataset0.json.gz']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(\"data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I'm putting these files into my ~/data, if it's different for u, change the path accordingly :)\n",
    "M6A_FILE_PATH = \"data/data.info.labelled\"\n",
    "DIRECT_RNA_SEQ_DATA_FILE_PATH = \"data/dataset0.json.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read m6a labels\n",
    "def read_m6A_labels(m6a_file_path):\n",
    "    m6a_df = pd.read_csv(m6a_file_path, sep=\",\")\n",
    "    m6a_df.columns = [\"gene_id\", \"transcript_id\", \"transcript_position\", \"label\"]\n",
    "    return m6a_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  transcript_position  label\n",
       "0  ENSG00000004059  ENST00000000233                  244      0\n",
       "1  ENSG00000004059  ENST00000000233                  261      0\n",
       "2  ENSG00000004059  ENST00000000233                  316      0\n",
       "3  ENSG00000004059  ENST00000000233                  332      0\n",
       "4  ENSG00000004059  ENST00000000233                  368      0"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m6a_labels_df = read_m6A_labels(M6A_FILE_PATH)\n",
    "m6a_labels_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2 Read direct rna seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_nucleotide_to_index(nucleotide):\n",
    "    nucleotide = nucleotide.lower()\n",
    "    if nucleotide == \"a\":\n",
    "        return 0\n",
    "    elif nucleotide == \"c\":\n",
    "        return 1\n",
    "    elif nucleotide == \"g\":\n",
    "        return 2\n",
    "    elif nucleotide == \"t\":\n",
    "        return 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_direct_rna_seq_data(data_path):\n",
    "    data = []\n",
    "    with gzip.open(data_path, 'rt') as f:\n",
    "        for line in f:\n",
    "            line_data = json.loads(line)\n",
    "            for transcript_id, position_data in line_data.items():\n",
    "                for transcript_position, combined_nucleotides_data in position_data.items():\n",
    "                    for combined_nucleotide, reads in combined_nucleotides_data.items():\n",
    "                        nucleotide_1, nucleotide_2, nucleotide_3, nucleotide_4, nucleotide_5, nucleotide_6, nucleotide_7 = combined_nucleotide\n",
    "                        nucleotide_1_idx, nucleotide_2_idx, nucleotide_3_idx, nucleotide_4_idx, nucleotide_5_idx, nucleotide_6_idx, nucleotide_7_idx = convert_nucleotide_to_index(nucleotide_1), convert_nucleotide_to_index(nucleotide_2), convert_nucleotide_to_index(nucleotide_3), convert_nucleotide_to_index(nucleotide_4), convert_nucleotide_to_index(nucleotide_5), convert_nucleotide_to_index(nucleotide_6), convert_nucleotide_to_index(nucleotide_7)\n",
    "                        for read_idx, read in enumerate(reads):\n",
    "                            data.append({\n",
    "                                'transcript_id': transcript_id,\n",
    "                                'transcript_position': int(transcript_position),\n",
    "                                'combined_nucleotide': combined_nucleotide,\n",
    "                                'nucleotide_1_index': nucleotide_1_idx,\n",
    "                                'nucleotide_2_index': nucleotide_2_idx,\n",
    "                                'nucleotide_3_index': nucleotide_3_idx,\n",
    "                                'nucleotide_4_index': nucleotide_4_idx,\n",
    "                                'nucleotide_5_index': nucleotide_5_idx,\n",
    "                                'nucleotide_6_index': nucleotide_6_idx,\n",
    "                                'nucleotide_7_index': nucleotide_7_idx,\n",
    "                                'read_id': read_idx,\n",
    "                                'x_1': read[0],\n",
    "                                'x_2': read[1],\n",
    "                                'x_3': read[2],\n",
    "                                'x_4': read[3],\n",
    "                                'x_5': read[4],\n",
    "                                'x_6': read[5],\n",
    "                                'x_7': read[6],\n",
    "                                'x_8': read[7],\n",
    "                                'x_9': read[8]\n",
    "                            })\n",
    "\n",
    "    df = pd.DataFrame(data)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq_data_df = read_direct_rna_seq_data(DIRECT_RNA_SEQ_DATA_FILE_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined_nucleotide</th>\n",
       "      <th>nucleotide_1_index</th>\n",
       "      <th>nucleotide_2_index</th>\n",
       "      <th>nucleotide_3_index</th>\n",
       "      <th>nucleotide_4_index</th>\n",
       "      <th>nucleotide_5_index</th>\n",
       "      <th>nucleotide_6_index</th>\n",
       "      <th>nucleotide_7_index</th>\n",
       "      <th>read_id</th>\n",
       "      <th>x_1</th>\n",
       "      <th>x_2</th>\n",
       "      <th>x_3</th>\n",
       "      <th>x_4</th>\n",
       "      <th>x_5</th>\n",
       "      <th>x_6</th>\n",
       "      <th>x_7</th>\n",
       "      <th>x_8</th>\n",
       "      <th>x_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00299</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00465</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.00664</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>0.00590</td>\n",
       "      <td>5.57</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>11.20</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>9.24</td>\n",
       "      <td>87.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>96</td>\n",
       "      <td>0.01200</td>\n",
       "      <td>3.73</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.02520</td>\n",
       "      <td>14.40</td>\n",
       "      <td>123.0</td>\n",
       "      <td>0.00510</td>\n",
       "      <td>4.16</td>\n",
       "      <td>81.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>0.01350</td>\n",
       "      <td>4.09</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.00540</td>\n",
       "      <td>5.71</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.00396</td>\n",
       "      <td>3.48</td>\n",
       "      <td>81.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>4.17</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.00973</td>\n",
       "      <td>5.68</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>1.60</td>\n",
       "      <td>81.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>99</td>\n",
       "      <td>0.00640</td>\n",
       "      <td>8.50</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>14.90</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>3.05</td>\n",
       "      <td>76.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  transcript_position combined_nucleotide  \\\n",
       "0   ENST00000000233                  244             AAGACCA   \n",
       "1   ENST00000000233                  244             AAGACCA   \n",
       "2   ENST00000000233                  244             AAGACCA   \n",
       "3   ENST00000000233                  244             AAGACCA   \n",
       "4   ENST00000000233                  244             AAGACCA   \n",
       "..              ...                  ...                 ...   \n",
       "95  ENST00000000233                  244             AAGACCA   \n",
       "96  ENST00000000233                  244             AAGACCA   \n",
       "97  ENST00000000233                  244             AAGACCA   \n",
       "98  ENST00000000233                  244             AAGACCA   \n",
       "99  ENST00000000233                  244             AAGACCA   \n",
       "\n",
       "    nucleotide_1_index  nucleotide_2_index  nucleotide_3_index  \\\n",
       "0                    0                   0                   2   \n",
       "1                    0                   0                   2   \n",
       "2                    0                   0                   2   \n",
       "3                    0                   0                   2   \n",
       "4                    0                   0                   2   \n",
       "..                 ...                 ...                 ...   \n",
       "95                   0                   0                   2   \n",
       "96                   0                   0                   2   \n",
       "97                   0                   0                   2   \n",
       "98                   0                   0                   2   \n",
       "99                   0                   0                   2   \n",
       "\n",
       "    nucleotide_4_index  nucleotide_5_index  nucleotide_6_index  \\\n",
       "0                    0                   1                   1   \n",
       "1                    0                   1                   1   \n",
       "2                    0                   1                   1   \n",
       "3                    0                   1                   1   \n",
       "4                    0                   1                   1   \n",
       "..                 ...                 ...                 ...   \n",
       "95                   0                   1                   1   \n",
       "96                   0                   1                   1   \n",
       "97                   0                   1                   1   \n",
       "98                   0                   1                   1   \n",
       "99                   0                   1                   1   \n",
       "\n",
       "    nucleotide_7_index  read_id      x_1   x_2    x_3      x_4    x_5    x_6  \\\n",
       "0                    0        0  0.00299  2.06  125.0  0.01770  10.40  122.0   \n",
       "1                    0        1  0.00631  2.53  125.0  0.00844   4.67  126.0   \n",
       "2                    0        2  0.00465  3.92  109.0  0.01360  12.00  124.0   \n",
       "3                    0        3  0.00398  2.06  125.0  0.00830   5.01  130.0   \n",
       "4                    0        4  0.00664  2.92  120.0  0.00266   3.94  129.0   \n",
       "..                 ...      ...      ...   ...    ...      ...    ...    ...   \n",
       "95                   0       95  0.00590  5.57  126.0  0.01200  11.20  127.0   \n",
       "96                   0       96  0.01200  3.73  124.0  0.02520  14.40  123.0   \n",
       "97                   0       97  0.01350  4.09  126.0  0.00540   5.71  127.0   \n",
       "98                   0       98  0.00830  4.17  121.0  0.00973   5.68  124.0   \n",
       "99                   0       99  0.00640  8.50  122.0  0.00531  14.90  117.0   \n",
       "\n",
       "        x_7    x_8   x_9  \n",
       "0   0.00930  10.90  84.1  \n",
       "1   0.01030   6.30  80.9  \n",
       "2   0.00498   2.13  79.6  \n",
       "3   0.00498   3.78  80.4  \n",
       "4   0.01300   7.15  82.2  \n",
       "..      ...    ...   ...  \n",
       "95  0.00564   9.24  87.3  \n",
       "96  0.00510   4.16  81.2  \n",
       "97  0.00396   3.48  81.4  \n",
       "98  0.00316   1.60  81.8  \n",
       "99  0.01030   3.05  76.1  \n",
       "\n",
       "[100 rows x 20 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rna_seq_data.head()\n",
    "rna_seq_data_df.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.3 Merge direct rna seq data with labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq_data_with_labels_df = rna_seq_data_df.merge(m6a_labels_df, on=[\"transcript_id\", \"transcript_position\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined_nucleotide</th>\n",
       "      <th>nucleotide_1_index</th>\n",
       "      <th>nucleotide_2_index</th>\n",
       "      <th>nucleotide_3_index</th>\n",
       "      <th>nucleotide_4_index</th>\n",
       "      <th>nucleotide_5_index</th>\n",
       "      <th>nucleotide_6_index</th>\n",
       "      <th>nucleotide_7_index</th>\n",
       "      <th>...</th>\n",
       "      <th>x_2</th>\n",
       "      <th>x_3</th>\n",
       "      <th>x_4</th>\n",
       "      <th>x_5</th>\n",
       "      <th>x_6</th>\n",
       "      <th>x_7</th>\n",
       "      <th>x_8</th>\n",
       "      <th>x_9</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.01770</td>\n",
       "      <td>10.40</td>\n",
       "      <td>122.0</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>10.90</td>\n",
       "      <td>84.1</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.53</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00844</td>\n",
       "      <td>4.67</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.01030</td>\n",
       "      <td>6.30</td>\n",
       "      <td>80.9</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.92</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.01360</td>\n",
       "      <td>12.00</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>2.13</td>\n",
       "      <td>79.6</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.06</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.00830</td>\n",
       "      <td>5.01</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>3.78</td>\n",
       "      <td>80.4</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.92</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>3.94</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.01300</td>\n",
       "      <td>7.15</td>\n",
       "      <td>82.2</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id  transcript_position combined_nucleotide  \\\n",
       "0  ENST00000000233                  244             AAGACCA   \n",
       "1  ENST00000000233                  244             AAGACCA   \n",
       "2  ENST00000000233                  244             AAGACCA   \n",
       "3  ENST00000000233                  244             AAGACCA   \n",
       "4  ENST00000000233                  244             AAGACCA   \n",
       "\n",
       "   nucleotide_1_index  nucleotide_2_index  nucleotide_3_index  \\\n",
       "0                   0                   0                   2   \n",
       "1                   0                   0                   2   \n",
       "2                   0                   0                   2   \n",
       "3                   0                   0                   2   \n",
       "4                   0                   0                   2   \n",
       "\n",
       "   nucleotide_4_index  nucleotide_5_index  nucleotide_6_index  \\\n",
       "0                   0                   1                   1   \n",
       "1                   0                   1                   1   \n",
       "2                   0                   1                   1   \n",
       "3                   0                   1                   1   \n",
       "4                   0                   1                   1   \n",
       "\n",
       "   nucleotide_7_index  ...   x_2    x_3      x_4    x_5    x_6      x_7  \\\n",
       "0                   0  ...  2.06  125.0  0.01770  10.40  122.0  0.00930   \n",
       "1                   0  ...  2.53  125.0  0.00844   4.67  126.0  0.01030   \n",
       "2                   0  ...  3.92  109.0  0.01360  12.00  124.0  0.00498   \n",
       "3                   0  ...  2.06  125.0  0.00830   5.01  130.0  0.00498   \n",
       "4                   0  ...  2.92  120.0  0.00266   3.94  129.0  0.01300   \n",
       "\n",
       "     x_8   x_9          gene_id  label  \n",
       "0  10.90  84.1  ENSG00000004059      0  \n",
       "1   6.30  80.9  ENSG00000004059      0  \n",
       "2   2.13  79.6  ENSG00000004059      0  \n",
       "3   3.78  80.4  ENSG00000004059      0  \n",
       "4   7.15  82.2  ENSG00000004059      0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq_data_with_labels_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with no labels: 0\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of rows with no labels: {len(rna_seq_data_with_labels_df[rna_seq_data_with_labels_df['label'].isnull()])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Strategy 1: Get Aggregate of all the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>combined_nucleotide</th>\n",
       "      <th>nucleotide_1_index</th>\n",
       "      <th>nucleotide_2_index</th>\n",
       "      <th>nucleotide_3_index</th>\n",
       "      <th>nucleotide_4_index</th>\n",
       "      <th>nucleotide_5_index</th>\n",
       "      <th>nucleotide_6_index</th>\n",
       "      <th>nucleotide_7_index</th>\n",
       "      <th>...</th>\n",
       "      <th>x_2</th>\n",
       "      <th>x_3</th>\n",
       "      <th>x_4</th>\n",
       "      <th>x_5</th>\n",
       "      <th>x_6</th>\n",
       "      <th>x_7</th>\n",
       "      <th>x_8</th>\n",
       "      <th>x_9</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006101</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>2242</td>\n",
       "      <td>GAAACAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.393768</td>\n",
       "      <td>104.312077</td>\n",
       "      <td>0.009123</td>\n",
       "      <td>3.659807</td>\n",
       "      <td>99.084541</td>\n",
       "      <td>0.007017</td>\n",
       "      <td>2.583527</td>\n",
       "      <td>88.815942</td>\n",
       "      <td>ENSG00000004478</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>35</td>\n",
       "      <td>AGGACAG</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>7.024259</td>\n",
       "      <td>116.333333</td>\n",
       "      <td>0.009167</td>\n",
       "      <td>7.397778</td>\n",
       "      <td>119.327778</td>\n",
       "      <td>0.006937</td>\n",
       "      <td>2.754444</td>\n",
       "      <td>83.496296</td>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>54</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3.403200</td>\n",
       "      <td>119.960000</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>6.730200</td>\n",
       "      <td>121.180000</td>\n",
       "      <td>0.006244</td>\n",
       "      <td>3.840600</td>\n",
       "      <td>83.824000</td>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>207</td>\n",
       "      <td>TTGACCA</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.563962</td>\n",
       "      <td>101.907547</td>\n",
       "      <td>0.005667</td>\n",
       "      <td>7.683396</td>\n",
       "      <td>116.830189</td>\n",
       "      <td>0.006573</td>\n",
       "      <td>4.287358</td>\n",
       "      <td>77.590566</td>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENST00000002165</td>\n",
       "      <td>367</td>\n",
       "      <td>TGGACCA</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.926346</td>\n",
       "      <td>116.057692</td>\n",
       "      <td>0.007552</td>\n",
       "      <td>8.440962</td>\n",
       "      <td>116.134615</td>\n",
       "      <td>0.006806</td>\n",
       "      <td>2.879038</td>\n",
       "      <td>78.242308</td>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id  transcript_position combined_nucleotide  \\\n",
       "0   ENST00000000233                  244             AAGACCA   \n",
       "1   ENST00000000233                  261             CAAACTG   \n",
       "2   ENST00000000233                  316             GAAACAG   \n",
       "3   ENST00000000233                  332             AGAACAT   \n",
       "4   ENST00000000233                  368             AGGACAA   \n",
       "..              ...                  ...                 ...   \n",
       "95  ENST00000001008                 2242             GAAACAA   \n",
       "96  ENST00000002165                   35             AGGACAG   \n",
       "97  ENST00000002165                   54             GGGACAT   \n",
       "98  ENST00000002165                  207             TTGACCA   \n",
       "99  ENST00000002165                  367             TGGACCA   \n",
       "\n",
       "    nucleotide_1_index  nucleotide_2_index  nucleotide_3_index  \\\n",
       "0                    0                   0                   2   \n",
       "1                    1                   0                   0   \n",
       "2                    2                   0                   0   \n",
       "3                    0                   2                   0   \n",
       "4                    0                   2                   2   \n",
       "..                 ...                 ...                 ...   \n",
       "95                   2                   0                   0   \n",
       "96                   0                   2                   2   \n",
       "97                   2                   2                   2   \n",
       "98                   3                   3                   2   \n",
       "99                   3                   2                   2   \n",
       "\n",
       "    nucleotide_4_index  nucleotide_5_index  nucleotide_6_index  \\\n",
       "0                    0                   1                   1   \n",
       "1                    0                   1                   3   \n",
       "2                    0                   1                   0   \n",
       "3                    0                   1                   0   \n",
       "4                    0                   1                   0   \n",
       "..                 ...                 ...                 ...   \n",
       "95                   0                   1                   0   \n",
       "96                   0                   1                   0   \n",
       "97                   0                   1                   0   \n",
       "98                   0                   1                   1   \n",
       "99                   0                   1                   1   \n",
       "\n",
       "    nucleotide_7_index  ...       x_2         x_3       x_4       x_5  \\\n",
       "0                    0  ...  4.223784  123.702703  0.009373  7.382162   \n",
       "1                    2  ...  3.216424  109.681395  0.006813  3.226535   \n",
       "2                    2  ...  2.940541  105.475676  0.007416  3.642703   \n",
       "3                    3  ...  6.476350  129.355000  0.008632  2.899200   \n",
       "4                    0  ...  6.415051  117.924242  0.011479  5.870303   \n",
       "..                 ...  ...       ...         ...       ...       ...   \n",
       "95                   0  ...  3.393768  104.312077  0.009123  3.659807   \n",
       "96                   2  ...  7.024259  116.333333  0.009167  7.397778   \n",
       "97                   3  ...  3.403200  119.960000  0.008390  6.730200   \n",
       "98                   0  ...  3.563962  101.907547  0.005667  7.683396   \n",
       "99                   0  ...  3.926346  116.057692  0.007552  8.440962   \n",
       "\n",
       "           x_6       x_7       x_8        x_9          gene_id label  \n",
       "0   125.913514  0.007345  4.386989  80.570270  ENSG00000004059     0  \n",
       "1   107.889535  0.007710  3.016599  94.290698  ENSG00000004059     0  \n",
       "2    98.947027  0.007555  2.087146  89.364324  ENSG00000004059     0  \n",
       "3    97.836500  0.006101  2.236520  89.154000  ENSG00000004059     0  \n",
       "4   121.954545  0.010019  4.260253  85.178788  ENSG00000004059     0  \n",
       "..         ...       ...       ...        ...              ...   ...  \n",
       "95   99.084541  0.007017  2.583527  88.815942  ENSG00000004478     0  \n",
       "96  119.327778  0.006937  2.754444  83.496296  ENSG00000001036     0  \n",
       "97  121.180000  0.006244  3.840600  83.824000  ENSG00000001036     0  \n",
       "98  116.830189  0.006573  4.287358  77.590566  ENSG00000001036     0  \n",
       "99  116.134615  0.006806  2.879038  78.242308  ENSG00000001036     0  \n",
       "\n",
       "[100 rows x 21 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregated_df = rna_seq_data_df.drop(columns=['read_id']).groupby(['transcript_id', 'transcript_position', 'combined_nucleotide','nucleotide_1_index', 'nucleotide_2_index', 'nucleotide_3_index', 'nucleotide_4_index', 'nucleotide_5_index', 'nucleotide_6_index', 'nucleotide_7_index']).mean().reset_index()\n",
    "aggregated_with_labels_df = aggregated_df.merge(m6a_labels_df, on=[\"transcript_id\", \"transcript_position\"], how=\"left\")\n",
    "aggregated_with_labels_df.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Train Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = aggregated_with_labels_df[['nucleotide_1_index', 'nucleotide_2_index', 'nucleotide_3_index', 'nucleotide_4_index', 'nucleotide_5_index', 'nucleotide_6_index', 'nucleotide_7_index','x_1', 'x_2', 'x_3', 'x_4', 'x_5', 'x_6', 'x_7', 'x_8', 'x_9']]\n",
    "labels = aggregated_with_labels_df['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.3, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X_train shape: (85286, 16)\n",
      "X_test shape: (36552, 16)\n",
      "y_train shape: (85286,)\n",
      "y_test shape: (36552,)\n"
     ]
    }
   ],
   "source": [
    "print(f\"X_train shape: {X_train.shape}\")\n",
    "print(f\"X_test shape: {X_test.shape}\")\n",
    "print(f\"y_train shape: {y_train.shape}\")\n",
    "print(f\"y_test shape: {y_test.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the numbers below, we can see that the labels data is extremely inbalanced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "label\n",
       "0    81427\n",
       "1     3859\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\ruirong\\Anaconda3\\Lib\\site-packages\\sklearn\\linear_model\\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    }
   ],
   "source": [
    "model = LogisticRegression(class_weight='balanced') # added 'balanced' to class weights to mitigate the effects of imbalanced dataset\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "y_pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open('y_pred.pkl', 'wb') as file:\n",
    "    pickle.dump(y_pred, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Evaluation\n",
    "\n",
    "Just going to use a basic balanced accuracy here for now. We can explore the usage of ROC next."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "report = classification_report(y_test, y_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Balanced Accuracy: 0.7579866692588655\n",
      "Classification Report:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.76      0.86     34936\n",
      "           1       0.13      0.76      0.22      1616\n",
      "\n",
      "    accuracy                           0.76     36552\n",
      "   macro avg       0.56      0.76      0.54     36552\n",
      "weighted avg       0.95      0.76      0.83     36552\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"Balanced Accuracy: {balanced_accuracy}\")\n",
    "print(\"Classification Report:\")\n",
    "print(report)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XGBoost Model\n",
    "\n",
    "we first try an XGBoost model to see its performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xgboost as xgb\n",
    "\n",
    "scale_pos_weight = len(y_train) / (sum(y_train == 1))\n",
    "\n",
    "# Create an XGBoost Classifier\n",
    "model = xgb.XGBClassifier(scale_pos_weight=scale_pos_weight, random_state=42)\n",
    "\n",
    "# Train the model\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions\n",
    "y_pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.91      0.95     34936\n",
      "           1       0.27      0.70      0.39      1616\n",
      "\n",
      "    accuracy                           0.90     36552\n",
      "   macro avg       0.63      0.81      0.67     36552\n",
      "weighted avg       0.95      0.90      0.92     36552\n",
      "\n",
      "0.8078679798080564\n"
     ]
    }
   ],
   "source": [
    "# Evaluate the model\n",
    "print(classification_report(y_test, y_pred))\n",
    "\n",
    "balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "\n",
    "print(balanced_accuracy)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we try gridsearch to optimize the parameters used during training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Parameters: {'subsample': 0.7999999999999999, 'reg_lambda': 0.1, 'reg_alpha': 1, 'n_estimators': 300, 'max_depth': 3, 'learning_rate': 0.21000000000000002, 'gamma': 0.1, 'colsample_bytree': 0.7999999999999999}\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.86      0.92     34936\n",
      "           1       0.20      0.78      0.32      1616\n",
      "\n",
      "    accuracy                           0.85     36552\n",
      "   macro avg       0.59      0.82      0.62     36552\n",
      "weighted avg       0.95      0.85      0.89     36552\n",
      "\n",
      "Balanced Accuracy: 0.8164444829243629\n"
     ]
    }
   ],
   "source": [
    "import xgboost as xgb\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "from sklearn.metrics import classification_report, balanced_accuracy_score\n",
    "import numpy as np\n",
    "\n",
    "# Define the parameter distribution\n",
    "param_dist = {\n",
    "    'n_estimators': [100, 200, 300],\n",
    "    'max_depth': [3, 5, 7, 9],\n",
    "    'learning_rate': np.arange(0.01, 0.3, 0.05),\n",
    "    'subsample': np.arange(0.5, 1.0, 0.1),\n",
    "    'colsample_bytree': np.arange(0.5, 1.0, 0.1),\n",
    "    'gamma': [0, 0.1, 0.2, 0.3],\n",
    "    'reg_alpha': [0, 0.01, 0.1, 1],\n",
    "    'reg_lambda': [0, 0.01, 0.1, 1]\n",
    "}\n",
    "\n",
    "# Create the XGBoost classifier\n",
    "xgb_model = xgb.XGBClassifier(scale_pos_weight=scale_pos_weight, random_state=42)\n",
    "\n",
    "# Create the RandomizedSearchCV object\n",
    "random_search = RandomizedSearchCV(estimator=xgb_model,\n",
    "                                   param_distributions=param_dist,\n",
    "                                   scoring='balanced_accuracy',\n",
    "                                   n_iter=100,  # Number of parameter settings to sample\n",
    "                                   cv=3,\n",
    "                                   n_jobs=-1,\n",
    "                                   random_state=42)\n",
    "\n",
    "# Fit the random search to the data\n",
    "random_search.fit(X_train, y_train)\n",
    "\n",
    "# Get the best parameters\n",
    "best_params = random_search.best_params_\n",
    "print(\"Best Parameters:\", best_params)\n",
    "\n",
    "# Use the best parameters to create a new model\n",
    "best_model = xgb.XGBClassifier(**best_params, scale_pos_weight=scale_pos_weight, random_state=42)\n",
    "\n",
    "# Train the best model\n",
    "best_model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions\n",
    "y_pred_best = best_model.predict(X_test)\n",
    "\n",
    "# Evaluate the best model\n",
    "print(classification_report(y_test, y_pred_best))\n",
    "balanced_accuracy_best = balanced_accuracy_score(y_test, y_pred_best)\n",
    "print(\"Balanced Accuracy:\", balanced_accuracy_best)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The performance is significantly higher than the initial logistics regression model, so i think this can be our benchmark model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
